libecho/src/BitReceiver.cc

```diff
@@ -0,0 +1,206 @@
+    sync_in.reserve(3 * win_size);
+    sync_lo_out.reserve(2 * win_size);
+    sync_hi_out.reserve(2 * win_size);
+
+    this->lo_ratio = static_cast<double>(lo_freq) / SAMPLE_RATE;
+    this->hi_ratio = static_cast<double>(hi_freq) / SAMPLE_RATE;
+    this->win_size = win_size;
+    this->mag_lim = mag_lim;
+}
+
+void BitReceiver::readSamples(int16_t *buffer, int len) {
+    int inc = 0;
+    int bytes = len * 2;
+    while (bytes > 0) {
+        int nread = input->readBytes((char *)buffer + inc, bytes);
+        inc += nread;
+        bytes -= nread;
+        input->waitForTick();
+    }
+}
+
+int BitReceiver::stepShift(double *buffer, int size) {
+    double sum = 0;
+    double delta;
+    double diff;
+    // TODO: this cutoff point is loosely dependent on the window size and the
+    // volume and noisiness of the microphone. I don't think there's a reasonable
+    // way to compute this, but who knows. For now it's just hardcoded.
+    double diff_min = 10;
+    int res = -1;
+
+    for (int i = 0; i < size; i++) {
+        sum += buffer[i];
+    }
+
+    for (int i = 0; i < size; i++) {
+        sum += buffer[i + size];
+        delta = buffer[i + size] - buffer[i];
+        if (delta < mag_lim)
+            continue;
+
+        diff = sum - 0.5 * ((size + 1) * (buffer[i + size] + buffer[i]));
+        diff = std::abs(diff);
+        if (diff_min > diff) {
+            diff_min = diff;
+            res = i;
+        }
+        sum -= buffer[i];
+    }
+
+    return res;
+}
+
+int BitReceiver::decodeBit(int16_t *window) {
+    double lo = dft(window, win_size, lo_ratio);
+    double hi = dft(window, win_size, hi_ratio);
+    return static_cast<int>(lo < hi);
+}
+
+int BitReceiver::receiveFirstTwoBits() {
+    int res = 0;
+    int lo_shift = -1;
+    int hi_shift = -1;
+    int shift = -1;
+
+    readSamples(sync_in.data(), win_size * 2);
+
+    qDebug() << "Waiting for message start";
+
+    do {
+        readSamples(sync_in.data() + (2 * win_size), win_size);
+        dftSlide(sync_in.data(), win_size, lo_ratio, sync_lo_out.data(), win_size * 2);
+        dftSlide(sync_in.data(), win_size, hi_ratio, sync_hi_out.data(), win_size * 2);
+        for (int i = 0; i < 2 * win_size; i++) {
+            sync_in[i] = sync_in[i + win_size];
+        }
+        lo_shift = stepShift(sync_lo_out.data(), win_size);
+        hi_shift = stepShift(sync_hi_out.data(), win_size);
+        if (lo_shift != -1) shift = lo_shift;
+        if (hi_shift != -1) shift = hi_shift;
+    } while (shift == -1);
+
+    qDebug() << "Message start detected!";
+
+    for (int i = 0; i < 2 * win_size - shift; i++) {
+        sync_in[i] = sync_in[i + win_size + shift];
+    }
+    readSamples(sync_in.data() + (2 * win_size - shift), shift);
+
+    res += decodeBit(sync_in.data());
+    res += decodeBit(sync_in.data() + win_size) << 1;
+
+    return res;
+}
+
+int BitReceiver::receiveFirst(uint8_t *buffer, int size) {
+    uint8_t byte = 0;
+
+    input->startStream();
```

Comment:
Maybe move those starts and stops to constructor/destructor as we discussed, so stream is always running?